willofsteel/async_client.py

```diff
@@ -0,0 +1,319 @@
+
+        Parameters
+        ----------
+        offer_type: :class:`Literal["buy", "sell"]`
+            The type of offer to retrieve.
+        item_id: :class:`str`
+            The ID of the item to retrieve offers for.
+
+        """
+        if offer_type not in ["buy", "sell"]:
+            raise InvalidInput("offer_type")
+        offers = []
+        params = {
+            "order_type": offer_type,
+            "item_type": item_id
+        }
+        data, status = await self.request(
+            "GET", "/market", headers=self.headers, params=params)
+        logging.debug(f"Got offer data for {item_id}: {str(data)}")
+        if status != 200:
+            parse_error(data["detail"])
+        for order_uuid, order_data in data["orders"].items():
+            offers.append(MarketOrder.from_response(order_uuid, order_data))
+        return offers
+
+    async def recruit_troop(self, unit_type: UnitType, amount: int, currency: Literal["gold", "silver"] = "gold") -> bool:
+        """
+        Recruit troops.
+
+        Parameters
+        ----------
+        unit_type: :class:`~willofsteel.types.UnitType`
+            The type of unit to recruit.
+        amount: :class:`int`
+            The amount of units to recruit.
+        currency: :class:`str`
+            The currency to use for recruitment. Defaults to gold.
+
+        Returns
+        -------
+        :class:`bool`
+
+        """
+        new_troop_name = unit_type.name.lower().replace(" ", "_").replace("'", "")
+        query_params = {
+            "troop": new_troop_name,
+            "amount": amount,
+            "currency": currency
+        }
+        data, status = await self.request("POST", "/recruit", self.headers, query_params)
+        if status == 200:
+            logging.debug("Troop recruitment was successful. Resp code: 200")
+            return True
+        else:
+            parse_error(data["detail"])
+            print(data["detail"])
+            print("This error was not automatically detected, please report this to the maintainers (or fix it yourself)!")
+
+    async def request(self, method: Literal["GET", "POST"], route: str, headers: dict, params: dict = None):
+        """
+        Send an HTTP request and return the response.
+
+        Parameters
+        ----------
+        method: :class:`str`
+            The HTTP method to use.
+        route: :class:`str`
+            The route to send the request to.
+        headers: :class:`dict`
+            The headers to send with the request.
+        params: :class:`dict`
+            The query parameters to send with the request. Defaults to None.
+
+        Returns
+        -------
+        :class:`dict` The json response from the api,
+        :class:`int` The status code of the response
+
+        """
+        url = BASE + route
+
+        if method not in ["GET", "POST"]:
+            return KeyError("Invalid Method")
+
+        async with aiohttp.ClientSession() as session:
```

Comment:
Huge blunder.

We do not want to create another session on every request, rather create a session on initialization and use it till the program is killed. 

Please try to accomplish this functionality asap